keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,445 @@
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a machine learning researcher I want to run a workload, comprising tens of
+pods that all need to run at the same time. Autoscaling the cluster and
+cheduling of all pods in this workload take a couple of minutes and other
+workloads scheduled within this interval can occupy the resources my workload
+could use otherwise, potentially forcing additional scale-up events. I would
+like to be sure that my workload, once admitted, can get resources allocated to
+all pods without an unnecessary delay.
+
+My use case be supported by adding `kueue.x-k8s.io/startup-requirements: QuotaAndPodsReady`
+annotation to the jobs I run.
+
+#### Story 2
+
+As an administrator of a batch processing in my organization I want to be able
+to configure kueue quotas for on-demand resources a little higher than what is
+available in the quiet period. However, I don't want to risk two workloads
+deadlock in case of issues with provisioning nodes to satisfy the configured
+quota. This could happen when the job corresponding to these workloads don't
+specify priorities or specify the same priority.
+
+My use case be supported by adding `kueue.x-k8s.io/startup-requirements: QuotaAndPodsReady`
+annotation to all jobs.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+To mitigate this issue we introduce introduce a timeout on reaching the `PodsReady`
+condition by a workload since its job start (see:
+[Timeout on reaching the PodsReady condition](#timeout-on-reaching-the-podsready-condition)).
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### PodsReady workload condition
+
+First, we introduce a new workload condition, called `PodsReady`, to indicate
+that the first batch of a workload's pods is scheduled, more precisely:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
```

Comment:
What about job.status.failed? What if pods fail quickly before we notice they were ready? And if we do count "ready in the past, but now terminated", maybe PodsStarted as a name?

Reply:
I think counting failed pods actually make sense. If the workload is a buggy one, we don't want to clog the queue waiting for it to have running or succeeded pods.

Reply:
Failed pods are counted already towards the `job.spec.backoffLimit` so a buggy workload would be eliminated based on exceeding the backoff limit. Seems good enough, let me add a sentence in this section.